torch/_functorch/partitioners.py

```diff
@@ -1016,11 +1026,50 @@ def default_partition(
     forward_node_names = OrderedSet(
         node.name for node in forward_only_graph.nodes if node.op != "output"
     )
+    order = {node: idx for idx, node in enumerate(joint_module.graph.nodes)}
     saved_values = []
     saved_sym_nodes = []
 
+    def is_mutated_later_in_fw(node):
+        if _has_tag_is_backward(node):
+            return False
+        tensor_arg_aliases = [
+            x
+            for x in node.args
+            if isinstance(x, fx.Node)
+            and "val" in x.meta
+            and isinstance(x.meta["val"], torch.Tensor)
+        ]
+        while len(tensor_arg_aliases) > 0:
+            a = tensor_arg_aliases.pop()
+            for u in a.users:
+                if not isinstance(u.target, torch._ops.OpOverload):
+                    continue
+                # If we witness a mutation on our node later, and that mutation is not "must be in backward",
+                # then our node needs to be computed in the forward (otherwise we will compute it on the mutated values)
+                if (
+                    # one of the args was mutated
+                    u.target._schema.is_mutable
+                    # and the mutation happens "later"
+                    and order[u] > order[node]
+                    # and the mutation happened during the forward
+                    and not (_has_tag_is_backward(u) or _has_tag_must_be_in_backward(u))
+                ):
+                    for idx, alias_info in enumerate(u.target._schema.arguments):
+                        if alias_info.is_write and u.args[idx] is a:
+                            return True
+                elif u.target.is_view:
+                    tensor_arg_aliases.append(u)
+        return False
```

Comment:
And so with a partitioner that exactly respects the original placement, I do not feel this function would be necessary.

Reply:
Yeah this is very reasonable. If you would rather I rewrite the default partitioner completely to do this I'm happy to. Otherwise, I'll land this PR to get something shippable and we can do a broader refactor.

If we want to re-write the default partitioner this way, we'll have to think about the AC interaction (if we trace AC we get AC support "for free" if we do this strategy, but if we don't trace AC then the default partitioner will still need a graph pass to duplicate recompute in the backward, similar to what Jeffrey's PR does)

Reply:
I basically want to do what Jeffreys PR does but in this pass lol